### ISSUE TITLE
Consider bringing back oneshot & level trigger

### ISSUE THREAD

**Issue by carllerche** [tokio-rs/mio#1049]
Oneshot was removed as part of the portable mio API. When we discussed it, the argument (made by me) was that you can simulate oneshot fairly easily with edge-triggered notifications. However, I have realized that there is a use case that is not possible to simulate with edge-triggered: polling blocking sockets.

There are cases where you want event notifications for blocking sockets (one case might be using stdout w/o switching it to non-blocking). To use edge-triggered notifications, an operation must be performed repeatedly until WOULDBLOCK is received. This is not possible with blocking sockets. Instead, with oneshot notifications, before attempting a blocking operation, one would register a blocking socket with oneshot, wait for the notification, then perform a *single* blocking operation. Since the notification was received, it is (pretty much) guaranteed that the operation will succeed w/o blocking.

While it could be brought back on a platform by platform basis (once we figure out how we want to do that). However, given the use case above, we should consider bringing oneshot back officially.

Thoughts @Thomasdezeeuw?

**Comment by Thomasdezeeuw** [comment]
I'm +1 for this. I have some tests in Gaea that use level trigger that I can port to Mio.

**Comment by Thomasdezeeuw** [comment]
After chatting with @carllerche we've decided to bring back both level and oneshot trigger, but only for Unix for now.

**Comment by carllerche** [comment]
I believe, the current strategy is to conditionally define `oneshot` and `level` on the `Interests` struct.

**Comment by Thomasdezeeuw** [comment]
Moving to v1.0

**Comment by nox** [comment]
Servo's ipc-channel also used level-triggered events and it seems non-trivial to switch it to edge-triggered ones.

https://github.com/servo/ipc-channel/blob/master/src/platform/unix/mod.rs#L461-L464

**Comment by Thomasdezeeuw** [comment]
@nox after a quick look I think you need to add a loop to `recv` (https://github.com/servo/ipc-channel/blob/82cbc216f78b866878b4226b628c863c90f1a109/src/platform/unix/mod.rs#L873-L967), or in `UnixCmsg::recv` (https://github.com/servo/ipc-channel/blob/82cbc216f78b866878b4226b628c863c90f1a109/src/platform/unix/mod.rs#L1040), calling `libc::recvmsg` until you get a [`WouldBlock`](https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.WouldBlock). Basically you'll need to drain the I/O source to ensure you'll get another event.

**Comment by Thomasdezeeuw** [comment]
Since there has been little to no interest in this I'm going to close this. If people still want this we can reopen.

**Comment by agrover** [comment]
I still want this (ONESHOT). Unix-only is fine. An existing solution I work on uses a single epoll instance polled from multiple threads. In order to use Mio instead, it would be best if an event only was indicated on one of the waiting threads, and not indicated on other waiting threads until the first waiting thread was finished and re-enabled the event.

**Comment by Thomasdezeeuw** [comment]
> In order to use Mio instead, it would be best if an event only was indicated on one of the waiting threads, and not indicated on other waiting threads until the first waiting thread was finished and re-enabled the event.

@agrover are edge-triggers good enough here? Or would e.g. getting a readable and then writable events also mess up the implementation?

**Comment by agrover** [comment]
> @agrover are edge-triggers good enough here? Or would e.g. getting a readable and then writable events also mess up the implementation?

I'm not understanding. epoll(7) says: "...the caller has the option to specify the EPOLLONESHOT flag, to tell epoll to disable the associated file descriptor after the receipt  of  an  event with epoll_wait(2)." That makes me think if ONESHOT is used, all events for that fd in the epoll interest list are disabled until rearmed. e.g. if a fd was readable and signalled, even if it became writable it would not signal again until rearmed. (If separate events for a single fd are desired, it discusses using dup(2) to duplicate the fd, and then register the duplicate fd with the epollset with different interest value.)

**Comment by Thomasdezeeuw** [comment]
> I'm not understanding. epoll(7) says: "...the caller has the option to specify the EPOLLONESHOT flag, to tell epoll to disable the associated file descriptor after the receipt of an event with epoll_wait(2)." That makes me think if ONESHOT is used, all events for that fd in the epoll interest list are disabled until rearmed. e.g. if a fd was readable and signalled, even if it became writable it would not signal again until rearmed. (If separate events for a single fd are desired, it discusses using dup(2) to duplicate the fd, and then register the duplicate fd with the epollset with different interest value.)

That is correct.

From your original comment

> In order to use Mio instead, it would be best if an event only was indicated on one of the waiting threads, and not indicated on other waiting threads until the first waiting thread was finished and re-enabled the event.

I was asking if edge triggers (what Mio currently uses) would be good enough to support your use case. Edge triggers will only trigger once I/O is ready and not again until its readiness is drained. For example when using a TCP socket, it will trigger readability once the peer socket send some data and only once all the pending data on the socket is read can it trigger another event. This should give a thread plenty of time to process the event without another thread attempt to process another event for the same fd.

**Comment by agrover** [comment]
Yes edge triggers work fine for my use case. But, if this work were to proceed and it wasn't too much extra work, I'd vote for supporting both oneshot and level trigger flags, if that's what you're getting at.

**Comment by Thomasdezeeuw** [comment]
> Yes edge triggers work fine for my use case. But, if this work were to proceed and it wasn't too much extra work, I'd vote for supporting both oneshot and level trigger flags, if that's what you're getting at, since that what it seems more people are interested in.

I'm a afraid it's not that easy to support oneshot triggers properly, that's the reason why it was removed in the first place.

**Comment by richardweinberger** [comment]
Can you please bring back level trigger? I can understand that you have dropped oneshot. But all target platforms should have support for level since `select` is specified in POSIX.1-2001.

**Comment by Thomasdezeeuw** [comment]
@richardweinberger what is your use case that doesn't work without oneshot triggers?

> Can you please bring back level trigger? I can understand that you have dropped oneshot. But all target platforms should have support for level since `select` is specified in POSIX.1-2001.

Note that we use kqueue/epoll and IOCP, which are different then select.

**Comment by richardweinberger** [comment]
I understand that. But this does not mean that you cannot support it.
Forcing everybody to use edge trigger is IMHO a bit harsh.
For simple applications level trigger is most of the time easier to use than edge.
Actually edge is the special case, most event libraries I know support only level and having edge support
is a extra feature.

Does mio support platforms which have no support for level (IOW just edge) trigger?

**Comment by Thomasdezeeuw** [comment]
> I understand that. But this does not mean that you cannot support it. Forcing everybody to use edge trigger is IMHO a bit harsh. For simple applications level trigger is most of the time easier to use than edge.

This is not really true. For example there is https://github.com/tokio-rs/mio/issues/1317, which could lead to a DOS attack quite easily in applications using level triggers. If you follow the following recipe for using edge triggers it's actually the easiest to use (and matches the `Future`s trait design quite nicely):

1. Attempt the I/O operation, e.g. a `read(2)` call. (`Future::poll`)
2. If it fails wait for an event and go to step 1. (`return Poll::Pending`)
3. If it succeeds your application can continue on as expected. (`return Poll::Ready`)

> Actually edge is the special case, most event libraries I know support only level and having edge support is a extra feature.

Don't really see how that is relevant.

> Does mio support platforms which have no support for level (IOW just edge) trigger?

I'm not sure about IOCP, but I think epoll and kqueue both support it.

But you still have given a valid use case.

**Comment by YorickPeterse** [comment]
@Thomasdezeeuw I'm currently using the [polling](https://github.com/smol-rs/polling) crate for https://inko-lang.org/. Prior to that I used a custom setup, which similar to polling used oneshot mode but used edge mode (polling defaults to level mode IIRC).

In my setup, a lightweight process (essentially a fiber) can perform certain non-blocking operations, such as reading from a socket. When the process tries to do so but encounters `E_WOULDBLOCK`, it (re)registers itself with the poller, then it suspends itself so another process can run. The poller is polled in a separate OS thread. Each event produced contains a token which is the pointer to the process. When such an event is produced, this poller thread wakes up the process that is waiting for the event (each event can only belong to one process).

Thanks to oneshot mode, I don't need to manually deregister the file descriptor _before_ rescheduling processes. I also think oneshot mode is a bit faster than doing so, as IIRC it essentially just flips a bit rather than (potentially) dropping a bunch of resources. 

So in my case I guess the question is this: 

Assuming for a moment that mio is able to support oneshot mode, would it actually be faster compared to manually de-registering a file descriptor after an edge event (and re-registering it later)? If not, I think oneshot mode doesn't need explicit support. If so, it would be nice to have.

**Comment by Thomasdezeeuw** [comment]
> Assuming for a moment that mio is able to support oneshot mode, would it actually be faster compared to manually de-registering a file descriptor after an edge event (and re-registering it later)? If not, I think oneshot mode doesn't need explicit support. If so, it would be nice to have.

I imagine it will be faster since your avoid a system call (deregistering), but I don't have any numbers for it. However support oneshot across OSs isn't easy, for example see https://github.com/tokio-rs/mio/issues/911#issuecomment-465619538 and https://github.com/tokio-rs/mio/issues/394.

**Comment by YorickPeterse** [comment]
@Thomasdezeeuw In that case it seems "emulating" oneshot-mode is the way to go. This does bring up a second question: can one call `reregister` on a file descriptor that has never been registered before? Right now I maintain some state per socket to determine if it had previously been registered or not, but I would like to get rid of that if possible.

**Comment by Thomasdezeeuw** [comment]
> @Thomasdezeeuw In that case it seems "emulating" oneshot-mode is the way to go.

How do you propose we do that? Because right now we don't keep any state for Unix file descriptor (sockets etc.).

> This does bring up a second question: can one call `reregister` on a file descriptor that has never been registered before (rather than `register` the first time, then use `reregister` from that point on)? Right now I maintain some state per socket to determine if it had previously been registered or not, but I would like to get rid of that if possible.

On Unix yes, on Windows you can't. So, for Mio the answer is no I'm afraid.

**Comment by YorickPeterse** [comment]
@Thomasdezeeuw Sorry, I wasn't clear: when I said "emulating" I meant that as a user of Mio you would deregister a descriptor yourself. So in my case I'd end up with something like this:

```
loop {
  for event in poller.poll() {
    poller.deregister(event.descriptor)
    reschedule(event.process)
  }
}
```

> On Unix yes, on Windows you can't. So, for Mio the answer is no I'm afraid.

Ah darn, that means I'll need to keep the flag around. It's not a big deal since it's just a single `bool` though, so thanks for answering :smiley:

**Comment by Thomasdezeeuw** [comment]
> With ET, the lib pushes the responsibility to the user to completely drain the Socket. I don't understand what mio gains from removing LT whereas there is an impact on the developers.

Level triggers didn't work reliably across platforms. So if you only tested on one OS it didn't always work on another. We found that really only edge triggers worked the same across platforms (or had differences we could fix).

Using edge triggers isn't that more difficult to use the level triggers though. You can use the following as a general pattern of how to use them:
1. Try the I/O operation, e.g. the read.
2. If 1. returned an `io::ErrorKind::WouldBlock` "error" `register` the I/O resource.
3. Once you get an `Event` for the I/O operation (e.g. readable for a `read(2)` operation) go to 1.
